src/TestResults/match-by-context.ts

```diff
@@ -0,0 +1,362 @@
+    },
+
+    status: assertion?.status ?? TestReconciliationState.Unknown,
+    shortMessage: assertion?.shortMessage ?? err,
+    terseMessage: assertion?.terseMessage,
+    lineNumberOfError:
+      assertion?.line >= test.start.line && assertion?.line <= test.end.line
+        ? assertion.line - 1
+        : test.end.line - 1,
+  };
+};
+
+/** mark all data and child containers unmatched */
+const toUnmatchedResults = (tContainer: ContainerNode<ItBlock>, err: string): TestResult[] => {
+  const results = tContainer.childData.map((n) => toMatchResult(n.data[0], err));
+  tContainer.childContainers.forEach((c) => results.push(...toUnmatchedResults(c, err)));
+  return results;
+};
+
+type MessageType = 'context-mismatch' | 'match-failed' | 'unusual-match' | 'duplicate-test-name';
+const makeWarning = (fileName: string, verbose: boolean) => (
+  messageType: MessageType,
+  contextType: ContextType,
+  source: ContainerNode<ItBlock> | DataNode<ItBlock>,
+  assertion?: ContainerNode<TestAssertionStatus> | DataNode<TestAssertionStatus>,
+  extraReason?: string
+): void => {
+  if (!verbose) {
+    return;
+  }
+
+  const output = (message: string): void =>
+    console.warn(`[${fileName}] ${message} \n source=`, source, `\n assertion=`, assertion);
+
+  const blockType = contextType === 'container' ? 'describe' : 'test';
+  switch (messageType) {
+    case 'context-mismatch':
+      output(
+        `!! context mismatched !! ${contextType} nodes are different under "${source.name}": `
+      );
+      break;
+    case 'match-failed': {
+      output(`!! match failed !! ${blockType}: "${source.name}": ${extraReason} `);
+      break;
+    }
+    case 'duplicate-test-name': {
+      output(
+        `duplicate test names in the same describe block is not recommanded and might not be matched reliably: ${blockType} ${source.name}`
+      );
+      break;
+    }
+    case 'unusual-match': {
+      output(`unusual match: ${extraReason} : ${blockType} ${source.name}: `);
+      break;
+    }
+  }
+};
+
+const hasSameStructure = <N1 extends BaseNodeType, N2 extends BaseNodeType>(
+  list1: N1[],
+  list2: N2[]
+): boolean => list1.length === list2.length;
+
+/**
+ * match tests container with assertion container by their context structure.
+ * @param fileName
+ * @param tContainer
+ * @param aContainer
+ * @param verbose turns on/off the debugging warning messages.
+ */
+
+export const matchByContext = (
+  fileName: string,
+  tContainer: ContainerNode<ItBlock>,
+  aContainer: ContainerNode<TestAssertionStatus>,
+  verbose = true
+): TestResult[] => {
+  const warning = makeWarning(fileName, verbose);
+
+  /**
+   * this is where the actual match occurred. Each test and assertion container pair will try
+   * to match both its child-data and child-container list recursively.
+   *
+   * @param _tContainer
+   * @param _aContainer
+   * @returns a list of TestResult collected from all its children (data + containers)
+   */
+  const _matchByContext = (
```

Comment:
I'm a bit unhappy about having three layers of functions, especially if one of those layers is just dedicated to caching the warning parameters. I would instead inline `_matchByContext`, `handleTestBlockMatch` and `handleDescribeBlockMatch`. Otherwise one has to first scroll down to see what the function is actually doing (without getting lost in indentation) and then scroll up to see how it is actually being done.
<details>
<summary>For comparison...</summary>

```ts
export const matchByContext = (
    fileName: string,
    tContainer: ContainerNode<ItBlock>,
    aContainer: ContainerNode<TestAssertionStatus>,
    verbose = true
): TestResult[] => {
    // in principle one could still use `const warning = makeWarning(fileName, verbose);` here

    // the match algorithm: first match by sequence if they have the same structure;
    // then fallback to simple name-based matching. Upon each test block, it invokes the
    // callback to process the matched results.
    const matchList = <N1 extends NodeType<ItBlock>, N2 extends NodeType<TestAssertionStatus>>(
        list1: N1[],
        list2: N2[],
        onResult: (n1: N1, n2: N2[]) => void
    ): void => {
        if (hasSameStructure(list1, list2)) {
            list1.forEach((n, idx) => onResult(n, [list2[idx]]));
        } else {
            warning(fileName, verbose, 'context-mismatch', 'container', tContainer, aContainer);
            list1.forEach((n) => {
                // duplicate names under the same layer is really illegal jest practice, they can not
                // be reliably resolved with name-based matching
                if (list1.filter((n1) => n1.name === n.name).length > 1) {
                    warning(fileName, verbose, 'duplicate-test-name', 'container', n);
                    onResult(n, []);
                } else {
                    const found = list2.filter((n2) => n2.name === n.name);
                    onResult(n, found);
                }
            });
        }
    };

    const matchResults: TestResult[] = [];

    matchList(tContainer.childData, aContainer.childData, (t, results) => {
        if (results.length !== 1) {
            matchResults.push(toMatchResult(t.only(), `found ${results.length} matched assertion(s)`));
            return;
        }
        const a = results[0];
        const itBlock = t.only();
        switch (a.data.length) {
            case 0:
                throw new TypeError(
                    `invalid state: assertion data should not be empty if it is a match!`
                );
            case 1: {
                const assertion = a.only();
                if (a.name !== t.name && !matchPos(itBlock, assertion)) {
                    warning(fileName, verbose, 'unusual-match', 'data', t, a, 'neither name nor line matched');
                }
                matchResults.push(toMatchResult(itBlock, assertion));
                break;
            }
            default: {
                // 1-to-many
                warning(fileName, verbose, 'unusual-match', 'data', t, a, '1-to-many match, jest.each perhaps?');

                // TODO: support multiple errorLine
                // until we support multiple errors, choose the first error assertion, if any
                const assertions =
                    a.data.find((assertion) => assertion.status === 'KnownFail') || a.first();
                matchResults.push(toMatchResult(itBlock, assertions));
            }
        }
    });

    matchList(tContainer.childContainers, aContainer.childContainers, (t, results) => {
        if (results.length === 1) {
            matchResults.push(...matchByContext(fileName, t, results[0], verbose));
        } else {
            warning(fileName, verbose, 'match-failed', 'container', t);
            // if we can't find corresponding container to match, the whole container will be considered unmatched
            matchResults.push(...toUnmatchedResults(t, `can not find matching assertion for block ${t.name}`));
        }
    });

    return matchResults;
};
```

</details>

Reply:
hmmm... ok, I am going to refactor this to flatten them out, it might make it more readable, but there will be redundant "setup" we have to do by not sharing the captured scope... maybe it is a worthy tradeoff...

Reply:
If one also flattens out `makeWarning` and supplies `fileName` and `verbose` each time when calling (like I did), there is no setup code remaining.

Reply:
I see what you mean, but I don't think by inline the functions back made it easier to read... that was the whole point for moving the handling logic out of the matchList call... anyway, I went the other way, now all of them are out of the match call, personally I found that easier to read than before...